GoogleMapsComponents/Maps/BaseListableEntity.cs

```diff
@@ -0,0 +1,106 @@
+﻿using Microsoft.JSInterop;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GoogleMapsComponents.Maps
+{
+    public abstract class BaseListableEntity<T> : IDisposable, IJsObjectRef
+        where T: BaseListableEntityOptions
+    {
+        protected readonly JsObjectRef _jsObjectRef;
+
+        public readonly Dictionary<string, List<MapEventListener>> EventListeners;
+
+        public Guid Guid => _jsObjectRef.Guid;
+
+        internal BaseListableEntity(JsObjectRef jsObjectRef)
+        {
+            _jsObjectRef = jsObjectRef;
+            EventListeners = new Dictionary<string, List<MapEventListener>>();
+        }
+
+        public void Dispose()
+        {
+            foreach (string key in EventListeners.Keys)
+            {
+                //Probably superfluous...
+                if (EventListeners[key] != null)
+                {
+                    foreach (MapEventListener eventListener in EventListeners[key])
+                    {
+                        eventListener.Dispose();
+                    }
+
+                    EventListeners[key].Clear();
+                }
+            }
+
+            EventListeners.Clear();
+            _jsObjectRef.Dispose();
+        }
+
+        public Task<Map> GetMap()
+        {
+            return _jsObjectRef.InvokeAsync<Map>(
+                "getMap");
+        }
+
+
+
+        /// <summary>
+        /// Renders the mao entity on the specified map or panorama. 
+        /// If map is set to null, the map entity will be removed.
+        /// </summary>
+        /// <param name="map"></param>
+        public async Task SetMap(Map map)
+        {
+            await _jsObjectRef.InvokeAsync(
+                   "setMap",
+                   map);
+
+            //_map = map;
+        }
+
+        public async Task<MapEventListener> AddListener(string eventName, Action handler)
+        {
+            JsObjectRef listenerRef = await _jsObjectRef.InvokeWithReturnedObjectRefAsync("addListener", eventName, handler);
+            MapEventListener eventListener = new MapEventListener(listenerRef);
+
+            if (!EventListeners.ContainsKey(eventName))
+            {
+                EventListeners.Add(eventName, new List<MapEventListener>());
+            }
+            EventListeners[eventName].Add(eventListener);
+
+            return eventListener;
+        }
+
+        public async Task<MapEventListener> AddListener<V>(string eventName, Action<V> handler)
+        {
+            JsObjectRef listenerRef = await _jsObjectRef.InvokeWithReturnedObjectRefAsync("addListener", eventName, handler);
```

Comment:
I will mention it only here. Why you dont use var. It is pleasure when you change function or anything and no need to make more 100 replaces to change type. I could agree that sometimes it help to understand purhose of variable, but many times variable name give purhose by itself. So better  name it more correclty. 
Also you dont use IEnumerable or ICollection. You will be swearing like hell one day if you do it in every solution.

Reply:
Same as before... "old school"
After a past of ASP (remote past ASP, not .Net) and "ulcer" caused by lack of strong typed implementation, I don't like "var", but sure, in some context it is not a sin :)